# Information Retrieval 1#
## Assignment 2: Retrieval models [100 points] ##

In this assignment you will get familiar with basic and advanced information retrieval concepts. You will implement different information retrieval ranking models and evaluate their performance.

We provide you with a Indri index. To query the index, you'll use a Python package ([pyndri](https://github.com/cvangysel/pyndri)) that allows easy access to the underlying document statistics.

For evaluation you'll use the [TREC Eval](https://github.com/usnistgov/trec_eval) utility, provided by the National Institute of Standards and Technology of the United States. TREC Eval is the de facto standard way to compute Information Retrieval measures and is frequently referenced in scientific papers.

This is a **groups-of-three assignment**, the deadline is **Wednesday, January 31st**. Code quality, informative comments and convincing analysis of the results will be considered when grading. Submission should be done through blackboard, questions can be asked on the course [Piazza](piazza.com/university_of_amsterdam/spring2018/52041inr6y/home).

### Technicalities (must-read!) ###

The assignment directory is organized as follows:
   * `./assignment.ipynb` (this file): the description of the assignment.
   * `./index/`: the index we prepared for you.
   * `./ap_88_90/`: directory with ground-truth and evaluation sets:
      * `qrel_test`: test query relevance collection (**test set**).
      * `qrel_validation`: validation query relevance collection (**validation set**).
      * `topics_title`: semicolon-separated file with query identifiers and terms.

You will need the following software packages (tested with Python 3.5 inside [Anaconda](https://conda.io/docs/user-guide/install/index.html)):
   * Python 3.5 and Jupyter
   * Indri + Pyndri (Follow the installation instructions [here](https://github.com/nickvosk/pyndri/blob/master/README.md))
   * gensim [link](https://radimrehurek.com/gensim/install.html)
   * TREC Eval [link](https://github.com/usnistgov/trec_eval)

### TREC Eval primer ###
The TREC Eval utility can be downloaded and compiled as follows:

    git clone https://github.com/usnistgov/trec_eval.git
    cd trec_eval
    make

TREC Eval computes evaluation scores given two files: ground-truth information regarding relevant documents, named *query relevance* or *qrel*, and a ranking of documents for a set of queries, referred to as a *run*. The *qrel* will be supplied by us and should not be changed. For every retrieval model (or combinations thereof) you will generate a run of the top-1000 documents for every query. The format of the *run* file is as follows:

    $query_identifier Q0 $document_identifier $rank_of_document_for_query $query_document_similarity $run_identifier
    
where
   * `$query_identifier` is the unique identifier corresponding to a query (usually this follows a sequential numbering).
   * `Q0` is a legacy field that you can ignore.
   * `$document_identifier` corresponds to the unique identifier of a document (e.g., APXXXXXXX where AP denotes the collection and the Xs correspond to a unique numerical identifier).
   * `$rank_of_document_for_query` denotes the rank of the document for the particular query. This field is ignored by TREC Eval and is only maintained for legacy support. The ranks are computed by TREC Eval itself using the `$query_document_similarity` field (see next). However, it remains good practice to correctly compute this field.
   * `$query_document_similarity` is a score indicating the similarity between query and document where a higher score denotes greater similarity.
   * `$run_identifier` is an identifier of the run. This field is for your own convenience and has no purpose beyond bookkeeping.
   
For example, say we have two queries: `Q1` and `Q2` and we rank three documents (`DOC1`, `DOC2`, `DOC3`). For query `Q1`, we find the following similarity scores `score(Q1, DOC1) = 1.0`, `score(Q1, DOC2) = 0.5`, `score(Q1, DOC3) = 0.75`; and for `Q2`: `score(Q2, DOC1) = -0.1`, `score(Q2, DOC2) = 1.25`, `score(Q1, DOC3) = 0.0`. We can generate run using the following snippet:

In [1]:
import logging
import sys
import os

def write_run(model_name, data, out_f,
              max_objects_per_query=sys.maxsize,
              skip_sorting=False):
    """
    Write a run to an output file.
    Parameters:
        - model_name: identifier of run.
        - data: dictionary mapping topic_id to object_assesments;
            object_assesments is an iterable (list or tuple) of
            (relevance, object_id) pairs.
            The object_assesments iterable is sorted by decreasing order.
        - out_f: output file stream.
        - max_objects_per_query: cut-off for number of objects per query.
    """
    for subject_id, object_assesments in data.items():
        if not object_assesments:
            logging.warning('Received empty ranking for %s; ignoring.',
                            subject_id)

            continue

        #Probe types, to make sure everything goes alright.
        assert isinstance(object_assesments[0][0], float) or \
             isinstance(object_assesments[0][0], np.float32)
        assert isinstance(object_assesments[0][1], str) or \
            isinstance(object_assesments[0][1], bytes)

        if not skip_sorting:
            object_assesments = sorted(object_assesments, reverse=True)

        if max_objects_per_query < sys.maxsize:
            object_assesments = object_assesments[:max_objects_per_query]

        if isinstance(subject_id, bytes):
            subject_id = subject_id.decode('utf8')

        for rank, (relevance, object_id) in enumerate(object_assesments):
            if isinstance(object_id, bytes):
                object_id = object_id.decode('utf8')

            out_f.write(
                '{subject} Q0 {object} {rank} {relevance} '
                '{model_name}\n'.format(
                    subject=subject_id,
                    object=object_id,
                    rank=rank + 1,
                    relevance=relevance,
                    model_name=model_name))
            
# The following writes the run to standard output.
# In your code, you should write the runs to local
# storage in order to pass them to trec_eval.
write_run(
    model_name='example',
    data={
        'Q1': ((1.0, 'DOC1'), (0.5, 'DOC2'), (0.75, 'DOC3')),
        'Q2': ((-0.1, 'DOC1'), (1.25, 'DOC2'), (0.0, 'DOC3')),
    },
    out_f=sys.stdout,
    max_objects_per_query=1000)

Q1 Q0 DOC1 1 1.0 example
Q1 Q0 DOC3 2 0.75 example
Q1 Q0 DOC2 3 0.5 example
Q2 Q0 DOC2 1 1.25 example
Q2 Q0 DOC3 2 0.0 example
Q2 Q0 DOC1 3 -0.1 example


Now, imagine that we know that `DOC1` is relevant and `DOC3` is non-relevant for `Q1`. In addition, for `Q2` we only know of the relevance of `DOC3`. The query relevance file looks like:

    Q1 0 DOC1 1
    Q1 0 DOC3 0
    Q2 0 DOC3 1
    
We store the run and qrel in files `example.run` and `example.qrel` respectively on disk. We can now use TREC Eval to compute evaluation measures. In this example, we're only interested in Mean Average Precision and we'll only show this below for brevity. However, TREC Eval outputs much more information such as NDCG, recall, precision, etc.

    $ trec_eval -m all_trec -q example.qrel example.run | grep -E "^map\s"
    > map                   	Q1	1.0000
    > map                   	Q2	0.5000
    > map                   	all	0.7500
    
Now that we've discussed the output format of rankings and how you can compute evaluation measures from these rankings, we'll now proceed with an overview of the indexing framework you'll use.

In [2]:
import pyndri

index = pyndri.Index('index/')

The loaded index can be used to access a collection of documents in an easy manner. We'll give you some examples to get some idea of what it can do, it is up to you to figure out how to use it for the remainder of the assignment.

First let's look at the number of documents, since Pyndri indexes the documents using incremental identifiers we can simply take the lowest index and the maximum document and consider the difference:

In [3]:
print("There are %d documents in this collection." % (index.maximum_document() - index.document_base()))

There are 164597 documents in this collection.


### Pyndri primer ###
For this assignment you will use [Pyndri](https://github.com/cvangysel/pyndri) [[1](https://arxiv.org/abs/1701.00749)], a python interface for [Indri](https://www.lemurproject.org/indri.php). We have indexed the document collection and you can query the index using Pyndri. We will start by giving you some examples of what Pyndri can do:

First we read the document collection index with Pyndri:

Let's take the first document out of the collection and take a look at it:

In [4]:
example_document = index.document(index.document_base())
print(example_document)

('AP890425-0001', (1360, 192, 363, 0, 880, 0, 200, 0, 894, 412, 92160, 3, 192, 0, 363, 34, 1441, 0, 174134, 0, 200, 0, 894, 412, 2652, 0, 810, 107, 49, 4903, 420, 0, 1, 48, 35, 489, 0, 35, 687, 192, 243, 0, 249311, 1877, 0, 1651, 1174, 0, 2701, 117, 412, 0, 810, 391, 245233, 1225, 5838, 16, 0, 233156, 3496, 0, 393, 17, 0, 2435, 4819, 930, 0, 0, 200, 0, 894, 0, 22, 398, 145, 0, 3, 271, 115, 0, 1176, 2777, 292, 0, 725, 192, 0, 0, 50046, 0, 1901, 1130, 0, 192, 0, 408, 0, 243779, 0, 0, 553, 192, 0, 363, 0, 3747, 0, 0, 0, 0, 1176, 0, 1239, 0, 0, 1115, 17, 0, 0, 585, 192, 1963, 0, 0, 412, 54356, 0, 773, 0, 0, 0, 192, 0, 0, 1130, 0, 363, 0, 545, 192, 0, 1174, 1901, 1130, 0, 4, 398, 145, 39, 0, 577, 0, 355, 0, 491, 0, 6025, 0, 0, 193156, 88, 34, 437, 0, 0, 1852, 0, 828, 0, 1588, 0, 0, 0, 2615, 0, 0, 107, 49, 420, 0, 0, 190, 7, 714, 2701, 0, 237, 192, 157, 0, 412, 34, 437, 0, 0, 200, 6025, 26, 0, 0, 0, 0, 363, 0, 22, 398, 145, 0, 200, 638, 126222, 6018, 0, 880, 0, 0, 161, 0, 0, 319, 894, 2701, 

Here we see a document consists of two things, a string representing the external document identifier and an integer list representing the identifiers of words that make up the document. Pyndri uses integer representations for words or terms, thus a token_id is an integer that represents a word whereas the token is the actual text of the word/term. Every id has a unique token and vice versa with the exception of stop words: words so common that there are uninformative, all of these receive the zero id.

To see what some ids and their matching tokens we take a look at the dictionary of the index:

In [5]:
token2id, id2token, _ = index.get_dictionary()
print(list(id2token.items())[:15])

[(1, 'new'), (2, 'percent'), (3, 'two'), (4, '1'), (5, 'people'), (6, 'million'), (7, '000'), (8, 'government'), (9, 'president'), (10, 'years'), (11, 'state'), (12, '2'), (13, 'states'), (14, 'three'), (15, 'time')]


Using this dictionary we can see the tokens for the (non-stop) words in our example document:

In [6]:
print([id2token[word_id] for word_id in example_document[1] if word_id > 0])

['52', 'students', 'arrested', 'takeover', 'university', 'massachusetts', 'building', 'fifty', 'two', 'students', 'arrested', 'tuesday', 'evening', 'occupying', 'university', 'massachusetts', 'building', 'overnight', 'protest', 'defense', 'department', 'funded', 'research', 'new', 'york', 'city', 'thousands', 'city', 'college', 'students', 'got', 'unscheduled', 'holiday', 'demonstrators', 'occupied', 'campus', 'administration', 'building', 'protest', 'possible', 'tuition', 'increases', 'prompting', 'officials', 'suspend', 'classes', '60', 'police', 'riot', 'gear', 'arrived', 'university', 'massachusetts', '5', 'p', 'm', 'two', 'hours', 'later', 'bus', 'drove', 'away', '29', 'students', 'camped', 'memorial', 'hall', 'students', 'charged', 'trespassing', '23', 'students', 'arrested', 'lying', 'bus', 'prevent', 'leaving', 'police', '300', 'students', 'stood', 'building', 'chanting', 'looking', 'students', 'hall', 'arrested', '35', 'students', 'occupied', 'memorial', 'hall', '1', 'p', 'm',

The reverse can also be done, say we want to look for news about the "University of Massachusetts", the tokens of that query can be converted to ids using the reverse dictionary:

In [7]:
query_tokens = index.tokenize("University of Massachusetts")
print("Query by tokens:", query_tokens)
query_id_tokens = [token2id.get(query_token,0) for query_token in query_tokens]
print("Query by ids with stopwords:", query_id_tokens)
query_id_tokens = [word_id for word_id in query_id_tokens if word_id > 0]
print("Query by ids without stopwords:", query_id_tokens)

Query by tokens: ['university', '', 'massachusetts']
Query by ids with stopwords: [200, 0, 894]
Query by ids without stopwords: [200, 894]


Naturally we can now match the document and query in the id space, let's see how often a word from the query occurs in our example document:

In [8]:
matching_words = sum([True for word_id in example_document[1] if word_id in query_id_tokens])
print("Document %s has %d word matches with query: \"%s\"." % (example_document[0], matching_words, ' '.join(query_tokens)))
print("Document %s and query \"%s\" have a %.01f%% overlap." % (example_document[0], ' '.join(query_tokens),matching_words/float(len(example_document[1]))*100))

Document AP890425-0001 has 13 word matches with query: "university  massachusetts".
Document AP890425-0001 and query "university  massachusetts" have a 2.5% overlap.


While this is certainly not everything Pyndri can do, it should give you an idea of how to use it. Please take a look at the [examples](https://github.com/cvangysel/pyndri) as it will help you a lot with this assignment.

**CAUTION**: Avoid printing out the whole index in this Notebook as it will generate a lot of output and is likely to corrupt the Notebook.

### Parsing the query file
You can parse the query file (`ap_88_89/topics_title`) using the following snippet:

In [9]:
import collections
import io
import logging
import sys

def parse_topics(file_or_files,
                 max_topics=sys.maxsize, delimiter=';'):
    assert max_topics >= 0 or max_topics is None

    topics = collections.OrderedDict()

    if not isinstance(file_or_files, list) and \
            not isinstance(file_or_files, tuple):
        if hasattr(file_or_files, '__iter__'):
            file_or_files = list(file_or_files)
        else:
            file_or_files = [file_or_files]

    for f in file_or_files:
        assert isinstance(f, io.IOBase)

        for line in f:
            assert(isinstance(line, str))

            line = line.strip()

            if not line:
                continue

            topic_id, terms = line.split(delimiter, 1)

            if topic_id in topics and (topics[topic_id] != terms):
                    logging.error('Duplicate topic "%s" (%s vs. %s).',
                                  topic_id,
                                  topics[topic_id],
                                  terms)

            topics[topic_id] = terms

            if max_topics > 0 and len(topics) >= max_topics:
                break

    return topics

with open('./ap_88_89/topics_title', 'r') as f_topics:
    print(parse_topics([f_topics]))

OrderedDict([('51', 'Airbus Subsidies'), ('52', 'South African Sanctions'), ('53', 'Leveraged Buyouts'), ('54', 'Satellite Launch Contracts'), ('55', 'Insider Trading'), ('56', 'Prime (Lending) Rate Moves, Predictions'), ('57', 'MCI'), ('58', 'Rail Strikes'), ('59', 'Weather Related Fatalities'), ('60', 'Merit-Pay vs. Seniority'), ('61', 'Israeli Role in Iran-Contra Affair'), ('62', "Military Coups D'etat"), ('63', 'Machine Translation'), ('64', 'Hostage-Taking'), ('65', 'Information Retrieval Systems'), ('66', 'Natural Language Processing'), ('67', 'Politically Motivated Civil Disturbances'), ('68', 'Health Hazards from Fine-Diameter Fibers'), ('69', 'Attempts to Revive the SALT II Treaty'), ('70', 'Surrogate Motherhood'), ('71', 'Border Incursions'), ('72', 'Demographic Shifts in the U.S.'), ('73', 'Demographic Shifts across National Boundaries'), ('74', 'Conflicting Policy'), ('75', 'Automation'), ('76', 'U.S. Constitution - Original Intent'), ('77', 'Poaching'), ('78', 'Greenpeace'

### Task 1: Implement and compare lexical IR methods [35 points] ### 

In this task you will implement a number of lexical methods for IR using the **Pyndri** framework. Then you will evaluate these methods on the dataset we have provided using **TREC Eval**.

Use the **Pyndri** framework to get statistics of the documents (term frequency, document frequency, collection frequency; **you are not allowed to use the query functionality of Pyndri**) and implement the following scoring methods in **Python**:

- [TF-IDF](http://nlp.stanford.edu/IR-book/html/htmledition/tf-idf-weighting-1.html) and 
- [BM25](http://nlp.stanford.edu/IR-book/html/htmledition/okapi-bm25-a-non-binary-model-1.html) with k1=1.2 and b=0.75. **[5 points]**
- Language models ([survey](https://drive.google.com/file/d/0B-zklbckv9CHc0c3b245UW90NE0/view))
    - Jelinek-Mercer (explore different values of 𝛌 in the range [0.1, 0.5, 0.9]). **[5 points]**
    - Dirichlet Prior (explore different values of 𝛍 [500, 1000, 1500]). **[5 points]**
    - Absolute discounting (explore different values of 𝛅 in the range [0.1, 0.5, 0.9]). **[5 points]**
    - [Positional Language Models](http://sifaka.cs.uiuc.edu/~ylv2/pub/sigir09-plm.pdf) define a language model for each position of a document, and score a document based on the scores of its PLMs. The PLM is estimated based on propagated counts of words within a document through a proximity-based density function, which both captures proximity heuristics and achieves an effect of “soft” passage retrieval. Implement the PLM, all five kernels, but only the Best position strategy to score documents. Use 𝛔 equal to 50, and Dirichlet smoothing with 𝛍 optimized on the validation set (decide how to optimize this value yourself and motivate your decision in the report). **[10 points]**
    
Implement the above methods and report evaluation measures (on the test set) using the hyper parameter values you optimized on the validation set (also report the values of the hyper parameters). Use TREC Eval to obtain the results and report on `NDCG@10`, Mean Average Precision (`MAP@1000`), `Precision@5` and `Recall@1000`.

For the language models, create plots showing `NDCG@10` with varying values of the parameters. You can do this by chaining small scripts using shell scripting (preferred) or execute trec_eval using Python's `subprocess`.

Compute significance of the results using a [two-tailed paired Student t-test](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_rel.html) **[5 points]**. Be wary of false rejection of the null hypothesis caused by the [multiple comparisons problem](https://en.wikipedia.org/wiki/Multiple_comparisons_problem). There are multiple ways to mitigate this problem and it is up to you to choose one.

Analyse the results by identifying specific queries where different methods succeed or fail and discuss possible reasons that cause these differences. This is *very important* in order to understand who the different retrieval functions behave.

**NOTE**: Don’t forget to use log computations in your calculations to avoid underflows. 

**IMPORTANT**: You should structure your code around the helper functions we provide below.

In [10]:
import time
start_time = time.time()
with open('./ap_88_89/topics_title', 'r') as f_topics:
    queries = parse_topics([f_topics])

index = pyndri.Index('index/')

num_documents = index.maximum_document() - index.document_base()

dictionary = pyndri.extract_dictionary(index)

tokenized_queries = {
    query_id: [dictionary.translate_token(token)
               for token in index.tokenize(query_string)
               if dictionary.has_token(token)]
    for query_id, query_string in queries.items()}

query_term_ids = set(
    query_term_id
    for query_term_ids in tokenized_queries.values()
    for query_term_id in query_term_ids)

print('Gathering statistics about', len(query_term_ids), 'terms.')

# inverted index creation.

document_lengths = {}
unique_terms_per_document = {}

inverted_index = collections.defaultdict(dict)
collection_frequencies = collections.defaultdict(int)

total_terms = 0

for int_doc_id in range(index.document_base(), index.maximum_document()):
    ext_doc_id, doc_token_ids = index.document(int_doc_id)

    document_bow = collections.Counter(
        token_id for token_id in doc_token_ids
        if token_id > 0)
    document_length = sum(document_bow.values())

    document_lengths[int_doc_id] = document_length
    total_terms += document_length

    unique_terms_per_document[int_doc_id] = len(document_bow)

    for query_term_id in query_term_ids:
        assert query_term_id is not None

        document_term_frequency = document_bow.get(query_term_id, 0)

        if document_term_frequency == 0:
            continue

        collection_frequencies[query_term_id] += document_term_frequency
        inverted_index[query_term_id][int_doc_id] = document_term_frequency


avg_doc_length = total_terms / num_documents
print('Inverted index creation took', time.time() - start_time, 'seconds.')

Gathering statistics about 456 terms.
Inverted index creation took 40.003620862960815 seconds.


In [11]:
test_query_ids = []
with open('./ap_88_89/qrel_test', 'r') as test_queries:
    for line in test_queries:
        query_id = line.split(' ')[0]
        if query_id not in test_query_ids:
            test_query_ids.append(query_id)

valid_query_ids = []
with open('./ap_88_89/qrel_validation', 'r') as valid_queries:
    for line in valid_queries:
        query_id = line.split(' ')[0]
        if query_id not in valid_query_ids:
            valid_query_ids.append(query_id)
            
print('We have ',len(test_query_ids),' and ',len(valid_query_ids),' query ids for the test and valid sets.')

We have  120  and  30  query ids for the test and valid sets.


In [12]:
def run_retrieval(model_name, score_fn):
    """
    Runs a retrieval method for all the queries and writes the TREC-friendly results in a file.
    
    :param model_name: the name of the model (a string)
    :param score_fn: the scoring function (a function - see below for an example) 
    """
    run_out_path = '{}.run'.format(model_name)

    if os.path.exists(run_out_path):
        print("File already exists")
        return
    
    retrieval_start_time = time.time()

    print('Retrieving using', model_name)
    data = {}
    count = 0
    # TODO: fill the data dictionary. 
    # The dictionary data should have the form: query_id --> (document_score, external_doc_id)
    if model_name == 'tf-idf' or model_name == 'BM25':
        for query_id, query in queries.items():
            score = 0.0
            query_word_ids = [token2id.get(query_token,0) for query_token in index.tokenize(query)] #get ids for all the words
            query_word_ids = [word_id for word_id in query_word_ids if word_id > 0] #remove stop words
            #print(query,query_word_ids)
            for int_doc_id in range(index.document_base(), index.maximum_document()):
                score = 0
                external_document_id,_ = index.document(int_doc_id) #Get the external doc id for this document
                for query_word in query_word_ids:
                    if int_doc_id in inverted_index[query_word].keys():
                        score += score_fn(int_doc_id,query_word,collection_frequencies[query_word])
                    else: 
                        score += 0.0
                    assert(isinstance(score,float))
                if query_id in data.keys():       
                    data[query_id].append((tuple([score,str(external_document_id)])))
                else:
                    data[query_id] = [((tuple([score,str(external_document_id)])))]
    elif model_name == 'JelinekMercer' or model_name == 'Dirichlet' or model_name == 'AbsoluteDiscounting':
        counter = 0
        for query_id, query in queries.items():
            if counter < 5:
                print(query_id)
                for int_doc_id in range(index.document_base(), index.maximum_document()):
                    external_document_id,_ = index.document(int_doc_id) #Get the external doc id for this document
                    score = score_fn(int_doc_id, query, collection_frequencies)
                    if query_id in data.keys():       
                        data[query_id].append((tuple([score,str(external_document_id)])))
                    else:
                        data[query_id] = [((tuple([score,str(external_document_id)])))]
                counter += 1
    else:
        print('Please give me a valid model_name!')
        
        #print(len(data[query_id]))
    with open(run_out_path, 'w') as f_out:
        write_run(
            model_name=model_name,
            data=data,
            out_f=f_out,
            max_objects_per_query=1000)

In [13]:
# The queries dict is an ordered dict with id's as keys and queris as values
print(queries['52'])

#Query term ids give for every unique term in the queries an id
print(list(query_term_ids)[:5])

# Inverted index gives for every query_term_id and doc_id the term frequency of that term in the document
#print(inverted_index[846])

# Collection frequencies gives for a collection of documents for every query term id the number of occurences of that term
print(collection_frequencies[8])

South African Sanctions
[1, 2053, 8, 4104, 6153]
119771


In [14]:
#Information Retrieval 1
import numpy as np
def tfidf(int_document_id, query_term_id, document_term_freq):
    """
    Scoring function for a document and a query term
    
    :param int_document_id: the document id
    :param query_token_id: the query term id (assuming you have split the query to tokens)
    :param document_term_freq: the document term frequency of the query term 
    """
    try:
        score = float(np.log(1+inverted_index[query_term_id][int_document_id]) * np.log(num_documents/document_term_freq))
    except:
        score = 0
    return score

# combining the two functions above: 
a = str(np.random.randint(0,100))

#run_retrieval('tfidf'+a, tfidf)

# TODO implement the rest of the retrieval functions

def BM25(int_document_id, query_term_id, document_term_freq): 
    k1 = 1.2
    b = 0.75
    # We ignore the sum including k3 since our queries are typically very short and therefore this part is not needed
    score = ((k1+1)*inverted_index[query_term_id][int_document_id])/(k1*((1-b)+b*(document_lengths[int_doc_id]/avg_doc_length) \
            + inverted_index[query_term_id][int_document_id]))*(num_documents/document_term_freq)
    
    return score

#run_retrieval('BM25'+a, BM25)

# TODO implement tools to help you with the analysis of the results.

In [15]:
test_query = queries['52']
test_document_id = 4

def JelinekMercer(int_document_id, query, collection_frequencies):
    lamb = 0.1 #We have to test 0.1/0.5/0.9
    query_prob = 1
    doc_length = len([i for i in index.document(int_document_id)[1] if i >0]) #get document length of non stopwords
    col_length = sum(collection_frequencies.values()) #get all counts of terms
    query_ids = [token2id.get(query_token,0) for query_token in index.tokenize(query)] #get ids for all the words
    query_ids = [word_id for word_id in query_ids if word_id > 0] #remove stop words  
    for word_id in query_ids:
        try:
            word_prob = (1-lamb)*(inverted_index[word_id][int_document_id]/doc_length)
        except:
            word_prob = 0
        word_prob += lamb*(collection_frequencies[word_id]/col_length)
        query_prob *= word_prob 
    return query_prob

print(JelinekMercer(test_document_id, test_query, collection_frequencies))

run_retrieval('JelinekMercer0.1', JelinekMercer)

1.962304247020076e-11
File already exists


Command for command line: ./trec_eval -m map -m P.5 -m ndcg -m recall.1000 qrel_test JelinekMercer.run

map                   	all	0.5604

P_5                   	all	0.7000

recall_1000           	all	0.9347

ndcg                  	all	0.8399

In [16]:
test_query = queries['52']
test_document_id = 1

def Dirichlet(int_document_id, query, collection_frequencies):
    mu = 500 #We have to test 500/1000/1500
    query_prob = 1
    doc_length = len([i for i in index.document(int_document_id)[1] if i >0]) #get document length of non stopwords
    col_length = sum(collection_frequencies.values()) #get all counts of terms
    query_ids = [token2id.get(query_token,0) for query_token in index.tokenize(query)] #get ids for all the words
    query_ids = [word_id for word_id in query_ids if word_id > 0] #remove stop words
    print(query,query_ids)    
    for word_id in query_ids:
        try:
            word_prob = (doc_length/(mu+doc_length))*(inverted_index[word_id][int_document_id]/doc_length)
        except:
            word_prob = 0
        word_prob += (mu/(mu+doc_length))*(collection_frequencies[word_id]/col_length)
        query_prob *= word_prob 
    return query_prob

print(Dirichlet(test_document_id, test_query, collection_frequencies))

run_retrieval('Dirichlet500', Dirichlet)

South African Sanctions [54, 846, 2208]
4.863368231857866e-09
File already exists


In [22]:
test_query = queries['52']
test_document_id = 1

def AbsoluteDiscounting(int_document_id, query, collection_frequencies):
    delta = 0.1 #We have to test 0.1/0.5/0.9
    query_prob = 1
    doc_length = len([i for i in index.document(int_document_id)[1] if i >0]) #get document length of non stopwords
    col_length = sum(collection_frequencies.values()) #get all counts of terms
    doc_unique_length = len([i for i in np.unique(example_document[int_document_id]) if i >0])
    sigma = delta*(doc_unique_length/doc_length)    
    query_ids = [token2id.get(query_token,0) for query_token in index.tokenize(query)] #get ids for all the words
    query_ids = [word_id for word_id in query_ids if word_id > 0] #remove stop words
    print(query,query_ids)    
    for word_id in query_ids:
        try:
            word_prob = max(inverted_index[word_id][int_document_id]-delta,0)/doc_length
        except:
            word_prob = 0
        word_prob += sigma*(collection_frequencies[word_id]/col_length)
        query_prob *= word_prob 
    return query_prob

print(AbsoluteDiscounting(test_document_id, test_query, collection_frequencies))

run_retrieval('AbsoluteDiscounting0.1', AbsoluteDiscounting)

South African Sanctions [54, 846, 2208]
3.981943903665423e-07
Retrieving using AbsoluteDiscounting0.1
Please give me a valid model_name!


In [23]:
from itertools import chain

def plm(int_document_id, query, position,kernel):
    sigma = 50
    score = 0
    tf_list = []
    denom = 0
    unique_query_terms = set(chain(*(query.split() for line in query if line)))
    unique_query_term_ids = [token2id[term.lower()] for term in unique_query_terms]
    _, doc_token_ids = index.document(int_document_id)

    for i in range(0,len(doc_token_ids)):
        if doc_token_ids[i] in unique_query_term_ids:
            tf_list.append(1)
        else:
            tf_list.append(0)
    
    if kernel == 'Gaussian':
        for i in range(0,len(tf_list)):
            score += tf_list[i]*np.exp(-(i-position)**2/(2*sigma**2))
            denom += 1*np.exp(-(i-position)**2/(2*sigma**2))
    
    if kernel == 'Triangle':
        for i in range(0,len(tf_list)):
            if sigma > abs(position-i):
                a = 0
            else:
                a = 1-(abs(position-i)/sigma)
        score += tf_list[i]*a
        denom += 1*np.exp(-(position-i)**2/(2*sigma**2))
    if kernel == 'Hamming':
        for i in range(0,len(tf_list)):
            if sigma > abs(position-i):
                a = 0
            else:
                a = 1/2*(1+np.cos(abs(position-i)*np.pi/sigma))
        score += tf_list[i]*a
        denom += 1*np.exp(-(position-i)**2/(2*sigma**2))
    if kernel == 'Circle':
        for i in range(0,len(tf_list)):
            if sigma > abs(position-i):
                a = 0
            else:
                a = np.sqrt(1-(abs(position-i)/sigma)**2)
        score += tf_list[i]*a
        denom += 1*np.exp(-(position-i)**2/(2*sigma**2))
    return score/denom

test_query = queries['52']
plm(9559,test_query,20,'Gaussian')

0.003154641556015966

### Task 2: Latent Semantic Models (LSMs) [15 points] ###

In this task you will experiment with applying distributional semantics methods ([LSI](http://lsa3.colorado.edu/papers/JASIS.lsi.90.pdf) **[5 points]** and [LDA](https://www.cs.princeton.edu/~blei/papers/BleiNgJordan2003.pdf) **[5 points]**) for retrieval.

You do not need to implement LSI or LDA on your own. Instead, you can use [gensim](http://radimrehurek.com/gensim/index.html). An example on how to integrate Pyndri with Gensim for word2vec can be found [here](https://github.com/cvangysel/pyndri/blob/master/examples/word2vec.py). For the remaining latent vector space models, you will need to implement connector classes (such as `IndriSentences`) by yourself.

In order to use a latent semantic model for retrieval, you need to:
   * build a representation of the query **q**,
   * build a representation of the document **d**,
   * calculate the similarity between **q** and **d** (e.g., cosine similarity, KL-divergence).
     
The exact implementation here depends on the latent semantic model you are using. 
   
Each of these LSMs come with various hyperparameters to tune. Make a choice on the parameters, and explicitly mention the reasons that led you to these decisions. You can use the validation set to optimize hyper parameters you see fit; motivate your decisions. In addition, mention clearly how the query/document representations were constructed for each LSM and explain your choices.

In this experiment, you will first obtain an initial top-1000 ranking for each query using TF-IDF in **Task 1**, and then re-rank the documents using the LSMs. Use TREC Eval to obtain the results and report on `NDCG@10`, Mean Average Precision (`MAP@1000`), `Precision@5` and `Recall@1000`.

Perform significance testing **[5 points]** (similar as in Task 1) in the class of semantic matching methods.

In [16]:
from gensim.models.keyedvectors import KeyedVectors
import pyndri.compat
import gensim
from gensim import corpora, models
import scipy.spatial.distance as ssd
import random
import copy

In [17]:
# Helper cell, that returns two dictionaries. These dictionaries can be used to convert 
# docIDs to external docIDs. Internal docIDS are 1,2,3,...,164597
# external docIDs are AP-XXXXXX-XXXX

ext2int_ids = {}
int2ext_ids = {}
        
for int_doc_id in range(index.document_base(), index.maximum_document()):
    ext_doc_id, _ = index.document(int_doc_id)
    ext2int_ids[ext_doc_id] = int_doc_id
    int2ext_ids[int_doc_id] = ext_doc_id
    
print (int2ext_ids[1])
print (ext2int_ids['AP890425-0001'])

AP890425-0001
1


## RETRIEVING THE MODEL

In [17]:
# Training is done by creating a very large BOW representation for all the documents in the collection
# These models are saved, so training has to be done only once. 

In [18]:
# A function that returns a BOW representation of the entire document set (160k documents)
# It loops through al the documents, and appends them to a list
def document_bow():
    dictionary = pyndri.extract_dictionary(index) 
    token2id, id2token, _ = index.get_dictionary() # Only id2token is necessary
    documents_list = [] # The list that all the documents will be appended to
    
    for i in range(1,num_documents+1):
        _ , doc = index.document(i)
        doc = [id2token[word_id] for word_id in doc if word_id > 0]
        documents_list.append(doc)
        
    bow_corpus = [dictionary.doc2bow(text) for text in documents_list]
    return bow_corpus, token2id, id2token

corpus, token2id, id2token = document_bow()

In [19]:
# Train LDA-model
num_topics = 20
lda20 = gensim.models.ldamodel.LdaModel(corpus=corpus, num_topics=num_topics, id2word = id2token, passes=10)
lda20.save('LDAmodels/LDA20')

In [24]:
# Train LSI-model

num_topics = 250
lsi250 = gensim.models.lsimodel.LsiModel(corpus=corpus, id2word=id2token, num_topics=num_topics)
lsi250.save('LSImodels/LSI250')
'''
num_topics = 500
lsi500 = gensim.models.lsimodel.LsiModel(corpus=corpus, id2word=id2token, num_topics=num_topics)
lsi100.save('LSImodels/LSI500')

num_topics = 1000
lsi1000 = gensim.models.lsimodel.LsiModel(corpus=corpus, id2word=id2token, num_topics=num_topics)
lsi1000.save('LSImodels/LSI1000')
'''

"\nnum_topics = 500\nlsi500 = gensim.models.lsimodel.LsiModel(corpus=corpus, id2word=id2token, num_topics=num_topics)\nlsi100.save('LSImodels/LSI500')\n\nnum_topics = 1000\nlsi1000 = gensim.models.lsimodel.LsiModel(corpus=corpus, id2word=id2token, num_topics=num_topics)\nlsi1000.save('LSImodels/LSI1000')\n"

## PREPROCESSING

In [19]:
# A function that returns a document given an external doc_id
# INPUT
# - an internal doc_id, a string, for instance: APXXXXX-XX
# RETURNS
# - an list of strings representing a text, for instance ['joris','is','de','beste']
def get_document(doc_id):
    int_doc_id = ext2int_ids[doc_id]
    _,text_ids = index.document(int_doc_id)
    return [id2token[word_id] for word_id in text_ids if word_id > 0]       

In [20]:
# Function that takes returns a list of indices, converting a text into indices
# INPUT
# - a list of strings, for instance ['python','wizard']
# RETURNS
# - a list of IDS, for instance [1,2] where 'python' maps to 1, and 'wizard' maps to 2
# according to the token2id dictionary
def text_to_ind(text):
    token_ids = [token2id[token] for token in text if token in token2id] 
    return token_ids

In [21]:
# A function that generates the topic vector from a text. This can be a document or a query
# INPUT
# - a text, a list of strings, for instance ['python','wizard']
# - a model, either LSI or LDA model
# RETURNS 
# - a topic scores vector, with scores
def topic_vector_from_text(text,model):
    random.shuffle(text) # randomly shuffle text, to really create a BOW
    bow_ids = text_to_ind(text)
    bow_ids_counter = collections.Counter(bow_ids)
    bow_list = [[key,value] for key,value in bow_ids_counter.items()]
    topics_scores = model[bow_list] # topics_scores is of type: [ (1,score1),(2,score2),etc,etc]
    
    topic_ids = [topic_id for topic_id,score in topics_scores]
    topic_scores = [score for topic_id,score in topics_scores]
    num_topics = model.num_topics
    
    scores = np.zeros(num_topics)
    scores[topic_ids] = topic_scores
    return scores

In [22]:
# Function that calculates the cosine similarity between two vectors
# INPUT
# - query_topic_vec, the query topic vector
# - doc_topic_vec, the document topic vector
# RETURNS
# - the cosine similarity between query_vec and doc_vec, a float
def check_similarity(query_topic_vec,doc_topic_vec):
    return ssd.cosine(query_topic_vec,doc_topic_vec)    

In [23]:
# Function that scores a list of document-ids based on a LSI or LDA model
# INPUT
# - a model, this can be either a LSI or LDA model
# - a query, this is a list of strings, in the form ['python','wizard']
# - a list of doc_ids, of the form [APXXXXX-XX,APYYYYY-YY, etc, etc]. This list is probably of length 1000
# RETURNS
# - a list of the 1000 documents, a list of form [(APYYYYY-YY,scoreY),(APXXXXX-XX,scoreX),(etc,score_etc),...],
# , where the first element of the list is the best query result, and the last element of the list is the worst 
def model_score(model,query,doc_ids):
    query_topic_vec = topic_vector_from_text(query,model)    
    score_list = []
    doc_ID_list = []
    
    for ext_doc_ID in doc_ids:
        doc = get_document(ext_doc_ID) # returns the document in text
        doc_topic_vec = topic_vector_from_text(doc,model)
        score = check_similarity(query_topic_vec,doc_topic_vec)
        
        # Ranking starts here. It loops through the score_list, inserting it in its correct position
        # A small value for score means a big match!
        if len(score_list) == 0: # first value is always the largest
            score_list.append(score)
            doc_ID_list.append(ext_doc_ID)
        else:
            for i in range(0,len(score_list)):
                if score_list[i] > score:
                    score_list.insert(i,score)
                    doc_ID_list.insert(i,ext_doc_ID)
                    break   
            if ext_doc_ID not in doc_ID_list:
                score_list.append(score)
                doc_ID_list.append(ext_doc_ID)
                                    
    assert len(score_list) == len(doc_ID_list) # test is everything went well
    ranking = []
    
    for i in range(0,len(score_list)):
        ranking.append((doc_ID_list[i],score_list[i]))
                    
    return ranking

In [24]:
# a function that reads the tfidf.run file
# INPUT
# - filename, a string
# RETURNS
# - a dictionary with takes a query_id (51,200) as value, and returns a list of a 1000 documents [APXXXX-XX,etc,...]
def read_tfidf_file(filename):
    
    return_dict = {}
    
    # Fill the dictionary with empty lists:
    # Queries go from 51 to 200
    #for i, qid in enumerate(queries):
     #   return_dict[int(qid)] = []
      #  print (qid)
        
    for qid in valid_query_ids:
        return_dict[int(qid)] = []
    
    with open(filename,'r') as fn:
        for line in fn:
            query_id = int(line.split()[0])
            ext_doc_ID = line.split()[2]
            
            temp_list = return_dict[query_id] # ugly solution this, but if its stupid but it works,
            temp_list.append(ext_doc_ID) # it aint stupid
            
            return_dict[query_id] = temp_list   
    return return_dict   

In [25]:
# A functions that appends data to a big data list
# INPUT
# - a list to_be_added, of the form [(APYYYYY-YY,scoreY),(APXXXXX-XX,scoreX),(etc,score_etc),...]
# - queryID, an integer
# - model_name a str with the model name
# - a data list of lists within a list, in which to add to_be_added
# - the lists within the list are of the form [queryID,'Q0',exc_doc_ID,rank,score,'modelname']
# RETURNS
# - data_list a list of list. Each list in the list is of the form [queryID,'Q0',ext_doc_ID,rank,score,model_name] 
def append_data(to_be_added,query_id,model_name,data_list):
    for i in range(0,len(to_be_added)):
        temp_list = []
        temp_list.append(query_id)
        temp_list.append('Q0')
        temp_list.append(to_be_added[i][0])
        temp_list.append(i+1) # rank
        temp_list.append(to_be_added[i][1])
        temp_list.append(model_name)
        data_list.append(temp_list)
    return data_list         

In [26]:
# Function that writes the data to a file
# INPUT
# - model_name, a string of the desired name of the output file
# - data, a list of list. Each list in the list is of the form [queryID,'Q0',ext_doc_ID,rank,score,model_name]
# RETURNS
# - writes data to a file, returns a file in the same folder as the .ipynb notebook
def write_model(model_name,data):
    with open(model_name,'w') as mn:
        for row in data:
            for term in row:
                mn.write(str(term) + ' ')
            mn.write('\n')
    mn.close()
    return

In [29]:
query_return_dict = read_tfidf_file('tfidf_valid.run')
#lsi_model = gensim.models.lsimodel.LsiModel.load('LSImodels/LSI250') 
lda_model = gensim.models.ldamodel.LdaModel.load('LDAmodels/LDA20')
model_name = 'LDA20.run'
data = []

for query,values in query_return_dict.items():
    query_in_text = queries[str(query)].lower().split()

    new_ranking = model_score(lda_model,query_in_text,query_return_dict[query])
    data = append_data(new_ranking,int(query),model_name,data)
    
write_model(model_name,data)
print ('done')

done


## Analysis

In [30]:
import os
from subprocess import Popen, PIPE
'''
print ('This is LDA10')
proc1 = Popen ('./evalLDA10.sh',shell=True,stdout = PIPE)
out,err = proc1.communicate()
print (out.decode('utf-8'))
'''
with open('logfile.txt','w') as file:

    for i in range(0,3):
        if i == 0:
            filename = 'LSI50'
        elif i == 1:
            filename = 'LSI100'
        else:
            filename = 'LSI250'
            
        file.write(filename + '\n')
        command = './eval'+filename + '.sh'
        proc = Popen (command,shell=True,stdout = PIPE)
        out,err = proc.communicate()
    
        for j in range (0,4):
            outline = out.decode('utf-8').split('\n')[j].split('\t')
            file.write(outline[0]+ ' ' + outline[-1] + '\n')
        
        file.write('\n')
        
    for i in range(0,2):
        if i == 0:
            filename = 'LDA10'
        else:
            filename = 'LDA20'
        
        file.write(filename + '\n')
        command = './eval'+filename + '.sh'
        proc = Popen (command,shell=True,stdout = PIPE)
        out,err = proc.communicate()
        
        for j in range (0,4):
            outline = out.decode('utf-8').split('\n')[j].split('\t')
            file.write(outline[0]+ ' ' + outline[-1] + '\n')
        
        file.write('\n')

file.close()
'''
print ()
print ('This is LSI100')
proc1 = Popen ('./evalLSI100.sh',shell=True,stdout = PIPE)
out,err = proc1.communicate()
print (out.decode('utf-8'))
print ()
print ('This is LSI250')
proc1 = Popen ('./evalLSI250.sh',shell=True,stdout = PIPE)
out,err = proc1.communicate()
print (out.decode('utf-8'))
'''

"\nprint ()\nprint ('This is LSI100')\nproc1 = Popen ('./evalLSI100.sh',shell=True,stdout = PIPE)\nout,err = proc1.communicate()\nprint (out.decode('utf-8'))\nprint ()\nprint ('This is LSI250')\nproc1 = Popen ('./evalLSI250.sh',shell=True,stdout = PIPE)\nout,err = proc1.communicate()\nprint (out.decode('utf-8'))\n"

In [205]:
'''
doc_list = ['AP890425-0001','AP890329-0221','AP880627-0093','AP880412-0268','AP880628-0170','AP890914-0250']
print (doc_list)
random.shuffle(doc_list)
print (doc_list)
model_score(lsi_model,['52','students'],doc_list)
'''       

"\ndoc_list = ['AP890425-0001','AP890329-0221','AP880627-0093','AP880412-0268','AP880628-0170','AP890914-0250']\nprint (doc_list)\nrandom.shuffle(doc_list)\nprint (doc_list)\nmodel_score(lsi_model,['52','students'],doc_list)\n"

### Task 3:  Word embeddings for ranking [20 points] (open-ended) ###

First create word embeddings on the corpus we provided using [word2vec](http://arxiv.org/abs/1411.2738) -- [gensim implementation](https://radimrehurek.com/gensim/models/word2vec.html). You should extract the indexed documents using pyndri and provide them to gensim for training a model (see example [here](https://github.com/nickvosk/pyndri/blob/master/examples/word2vec.py)).
   
This is an open-ended task. It is left up you to decide how you will combine word embeddings to derive query and document representations. Note that since we provide the implementation for training word2vec, you will be graded based on your creativity on combining word embeddings for building query and document representations.

Note: If you want to experiment with pre-trained word embeddings on a different corpus, you can use the word embeddings we provide alongside the assignment (./data/reduced_vectors_google.txt.tar.gz). These are the [google word2vec word embeddings](https://code.google.com/archive/p/word2vec/), reduced to only the words that appear in the document collection we use in this assignment.

In [6]:
import copy
import gensim
import logging
import pyndri
import pyndri.compat
import sys

dictionary = pyndri.extract_dictionary(index)
sentences = pyndri.compat.IndriSentences(index, dictionary)

word2vec = gensim.models.Word2Vec(sentences, min_count=1)
word2vec.save('W2V/w2v')


In [75]:
from nltk.stem.porter import PorterStemmer
# A function that returns a vector representation of a word
# INPUT
# - a word, a string, like 'joris'
# - word_vectors, a dictionary which returns the vector of a word
# CODE
# - returns the vector representation of a word. If this is not avalaible, it stems the word and finds the vector representation
# RETURNS
# - the vector representation of the (stemmed) word
def get_vector_for_word(word,word_vectors):
    try:
        embedding = word_vectors[str(word)]
        return embedding
    except KeyError:
        stemmed_word = PorterStemmer().stem(str(word))
        return word_vectors[stemmed_word]
#get_vector_for_word('panamas',word_vectors)

In [78]:
test_doc = get_document(int2ext_ids[3])

print (test_doc)

for word in test_doc:
    print (len(get_vector_for_word(word,word_vectors)))






['government', 'mandates', 'seatbelts', 'drivers', 'passengers', 'car', 'mad', 'country', 'buckle', 'beginning', 'thursday', 'keep', 'bambini', 'back', 'seat', 'government', 'announced', 'communique', 'premier', 'ciriaco', 'de', 'mitas', 'office', 'monday', 'people', 'caught', 'seat', 'belts', 'subject', 'fines', 'ranging', '12', '000', 'lire', '9', '50', '000', 'lire', '37', 'regulations', 'apply', 'pregnant', 'women', 'new', 'rules', 'require', 'babies', 'toddlers', 'car', 'seats', 'relegate', 'children', 'age', '12', 'back', 'seat', 'regulations', 'drawn', 'ministries', 'interior', 'public', 'works', 'transport', 'health']
300
300
300
300
300
300
300
300
300
300
300
300
300
300
300
300
300
300
300


KeyError: "word 'ciriaco' not in vocabulary"

In [63]:
# Function that loads the word vectors from the 'reduced_vectors_google.txt' file
# RETURNS
# - word_vectors
def load_word2vec():
    word_vectors = KeyedVectors.load_word2vec_format('reduced_vectors_google.txt', binary=False)
    
    return word_vectors

word_vectors = load_word2vec()

300


In [64]:
print (len(word_vectors['panama']))

300


senators 300
wimp 300
panama 300
united 300
states 300
wimp 300
fraud 300


KeyError: "word 'panamas' not in vocabulary"

### Task 4: Learning to rank (LTR) [15 points] (open-ended) ###

In this task you will get an introduction into learning to rank for information retrieval.

You can explore different ways for devising features for the model. Obviously, you can use the retrieval methods you implemented in Task 1, Task 2 and Task 3 as features. Think about other features you can use (e.g. query/document length). Creativity on devising new features and providing motivation for them will be taken into account when grading.

For every query, first create a document candidate set using the top-1000 documents using TF-IDF, and subsequently compute features given a query and a document. Note that the feature values of different retrieval methods are likely to be distributed differently.

You are adviced to start some pointwise learning to rank algorithm e.g. logistic regression, implemented in [scikit-learn](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).
Train your LTR model using 10-fold cross validation on the test set. More advanced learning to rank algorithms will be appreciated when grading.

### Task 4: Write a report [15 points; instant FAIL if not provided] ###

The report should be a PDF file created using the [sigconf ACM template](https://www.acm.org/publications/proceedings-template) and will determine a significant part of your grade.

   * It should explain what you have implemented, motivate your experiments and detail what you expect to learn from them. **[10 points]**
   * Lastly, provide a convincing analysis of your results and conclude the report accordingly. **[10 points]**
      * Do all methods perform similarly on all queries? Why?
      * Is there a single retrieval model that outperforms all other retrieval models (i.e., silver bullet)?
      * ...

**Hand in the report and your self-contained implementation source files.** Only send us the files that matter, organized in a well-documented zip/tgz file with clear instructions on how to reproduce your results. That is, we want to be able to regenerate all your results with minimal effort. You can assume that the index and ground-truth information is present in the same file structure as the one we have provided.
